common/src/main/java/io/pravega/common/util/btree/BTreePage.java

```diff
@@ -0,0 +1,788 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util.btree;
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.util.BitConverter;
+import io.pravega.common.util.ByteArraySegment;
+import io.pravega.common.util.IllegalDataFormatException;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import javax.annotation.concurrent.NotThreadSafe;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.val;
+
+/**
+ * B+Tree Page containing raw data. Wraps around a ByteArraySegment and formats it using a special layout.
+ *
+ * Format: Header|Data|Footer
+ * * Header: FormatVersion(1)|Flags(1)|Id(4)|Count(4)
+ * * Data: List{Key(KL)|Value(VL)}
+ * * Footer: Id(4)
+ *
+ * The Header contains:
+ * * The current format version
+ * * A set of flags that apply to this Page. Currently the only one used is to identify if it is an Index or Leaf page.
+ * * A randomly generated Page Identifier.
+ * * The number of items in the Page.
+ *
+ * The Data contains:
+ * * A list of Keys and Values, sorted by Key (using ByteArrayComparator). The length of this list is defined in the Header.
+ *
+ * The Footer contains:
+ * * The same Page Identifier as in the Header. When wrapping an existing ByteArraySegment, this value is matched to the
+ * one in the Header to ensure the Page was loaded correctly.
+ *
+ */
+@NotThreadSafe
+class BTreePage {
```

Comment:
I think we're going to want to move this off-heap if at all possible. 
What's the feasibility of having the payloads here be backed by Rocks?

Reply:
This doesn't stay in memory! It is loaded from external storage, we modify it, then we dump it back and get rid of it from heap. This B+Tree implementation is for an external data structure, it only keeps a couple of pointers in memory.